R/rectangle.R

```diff
@@ -167,104 +222,461 @@ hoist <- function(.data, .col, ..., .remove = TRUE, .simplify = TRUE, .ptype = l
+
+  for (i in seq_along(cols)) {
+    col <- cols[[i]]
+
+    data[[col]] <- df_simplify(
+      x = data[[col]],
+      ptype = ptype,
+      transform = transform,
+      simplify = simplify
+    )
+  }
+
+  unpack(data, all_of(cols), names_repair = names_repair)
+}
+
+# Converts a column of any type to a `list_of<tbl>`
+col_to_wide <- function(col, name, strict, names_sep) {
+  if (is.null(col)) {
+    abort(glue("Invalid `NULL` column detected for column `{name}`."))
+  }
+
+  if (!vec_is_list(col)) {
+    ptype <- vec_ptype(col)
+    col <- tidyr_chop(col)
+    col <- new_list_of(col, ptype = ptype)
+  }
+
+  # If we don't have a list_of, then a `NULL` `col_ptype` will get converted to
+  # a 1 row, 0 col tibble for `elt_ptype`
+  col_ptype <- list_of_ptype(col)
+  elt_ptype <- elt_to_wide(col_ptype, name = name, strict = strict, names_sep = names_sep)
+  elt_ptype <- vec_ptype(elt_ptype)
+
+  # Avoid expensive dispatch from `[[.list_of`
+  out <- tidyr_new_list(col)
+  out <- map(out, elt_to_wide, name = name, strict = strict, names_sep = names_sep)
+
+  # In the sole case of a list_of<data_frame>, we can be sure that the
+  # elements of `out` will all be of the same type. Otherwise,
+  # - If `col` isn't a list-of, we don't know the element type.
+  # - If `col` is a list-of but not one with a data frame ptype, then we
+  #   don't actually know if all elements have the same ptype, because the
+  #   number of resulting columns per element depends on that element's size.
+  has_identical_elements <- is_list_of(col) && is.data.frame(col_ptype)
+
+  if (has_identical_elements) {
+    ptype <- elt_ptype
+  } else {
+    ptype <- vec_ptype_common(elt_ptype, !!!out)
+    out <- vec_cast_common(!!!out, .to = ptype)
+  }
+
+  out <- new_list_of(out, ptype = ptype)
+
+  out
+}
+
+# Convert a list element to a wide tibble with:
+# - 1 row
+# - N cols, where `N = vec_size(x)`
+# - Column names come from `vec_names(x)`
+# - Data frames are treated specially. They are treated like heterogeneous lists
+#   where we know the type of each list element.
+# - When !strict, lists are treated specially as well. Any typed empty elements
+#   are replaced with `NULL` so their type doesn't interfere with the final
+#   common type. This is extremely useful for JSON data,
+#   where round tripping a typed empty cell results in an empty `list()` that
+#   won't be able to combine with other typed non-empty cells. However, it
+#   can create inconsistencies with vctrs typing rules.
+elt_to_wide <- function(x, name, strict, names_sep) {
+  if (is.null(x)) {
+    x <- list()
+  }
+
+  if (!vec_is(x)) {
+    abort(glue("Column `{name}` must contain a list of vectors."))
+  }
+
+  if (is.data.frame(x)) {
+    # Extremely special case for data.frames,
+    # which we want to treat like lists where we know the type of each element
+    x <- tidyr_new_list(x)
+    ptypes <- map(x, vec_ptype)
+    x <- tidyr_chop(x)
+    x <- map2(x, ptypes, new_list_of)
```

Comment:
I actually don't think the data frame behavior of `unnest_wider()` completely follows the semantics that I would expect. But what you have implemented is probably the most practically useful thing for it to do.

Here is what I would have expected (yes, it is weird):

``` r
library(tibble)

df <- tibble(x = list(tibble(a = 1:2, b = 2:3, c = 3:4)))
df
#> # A tibble: 1 × 1
#>   x               
#>   <list>          
#> 1 <tibble [2 × 3]>

df$x
#> [[1]]
#> # A tibble: 2 × 3
#>       a     b     c
#>   <int> <int> <int>
#> 1     1     2     3
#> 2     2     3     4

# unnest_wider(df, x)
# - The size of each element of `df$x` determines the number of columns.
#   2 rows (size) = 2 new columns
# - The `vec_names()` of the element determine the column names, but tibbles
#   don't use rownames...
tibble(...1 = df$x[[1]][1,], ...2 = df$x[[1]][2,])
#> # A tibble: 1 × 2
#>   ...1$a    $b    $c ...2$a    $b    $c
#>    <int> <int> <int>  <int> <int> <int>
#> 1      1     2     3      2     3     4

# This is more useful if you had a data frame with row names
df <- tibble(x = list(data.frame(a = 1:2, b = 2:3, c = 3:4, row.names = c("r1", "r2"))))
df$x
#> [[1]]
#>    a b c
#> r1 1 2 3
#> r2 2 3 4

# Then you'd get:
# unnest_wider(df, x)
tibble(r1 = df$x[[1]][1,], r2 = df$x[[1]][2,])
#> # A tibble: 1 × 2
#>    r1$a    $b    $c  r2$a    $b    $c
#>   <int> <int> <int> <int> <int> <int>
#> 1     1     2     3     2     3     4
```

<sup>Created on 2021-11-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>